File: scripts/02-data_cleaning.R
#### Preamble ####
# Purpose: Cleans the raw ambulance dispatch response data
# Author: Muhammad Abdullah Motasim
# Date: 22 September 2024
# Contact: abdullah.motasim@mail.utoronto.ca
# License: MIT
# Pre-requisites: tidyverse, dplyr, raw_ambulance_response_data.csv
# Any other information needed? None

#### Workspace setup ####
library(tidyverse)
library(dplyr)

#### Clean data ####
raw_data <- read_csv("data/raw_data/raw_ambulance_response_data.csv")

cleaned_data <- suppressWarnings(
  raw_data[,2:5] |>
  janitor::clean_names() |>
  # Remove columns which are not used in analysis
  # select(-forward_sortation_area, -field_name, -desctiption_definition, -comments_examples) |>
    
  na.omit() |>
  # Convert date and time variables
  separate(col = dispatch_time,
           into = c("date", "time"),
           sep = " ") |> 
  # Separate date variable 
  separate(col = date,
           into = c("year", "month", "day"),
           sep = "-") |>
  # Separate time variable
  separate(col = time,
           into = c("hour", "minute", "second"),
           sep = ":") |>
  # Ensure we keep values which are at time 00:00:00
  mutate(
    hour = ifelse(is.na(hour), "00", hour),
    minute = ifelse(is.na(minute), "00", minute),
    second = ifelse(is.na(second), "00", second)) |>
  # create a new date column
  mutate(dispatch_time = lubridate::ymd_hms(paste(year, month, day, hour, minute, second, sep = "-"))) |>
  # Remove unneeded columns which describe date
  select(-hour, -minute, -second, -day, -month) |>
  #Reorder columns
  select(dispatch_time, incident_type, priority_number, units_arrived_at_scene, year))

#### Test data ####

# Test for no negative number of dispatchers
paste("Negative Number of Dispatchers Test: ", cleaned_data$units_arrived_at_scene |> min() < 0)

# test for dates within range specified
years <- c(2017, 2018, 2019, 2020, 2021, 2022)
paste("Dispatch Time out-of-range Test: ", any(!cleaned_data$year %in% years))

# test for priority level meets requirements
priority_levels <- c(1, 3, 4, 5, 9, 11, 12, 13, 14)
paste("Priority Level Outside of Range Test: ", any(!cleaned_data$priority_number %in% priority_levels))

# test for missing values
any_NA_values <- all(is.na(cleaned_data$dispatch_time))
any_NA_values <- all(is.na(cleaned_data$incident_type))
any_NA_values <- all(is.na(cleaned_data$priority_number))
any_NA_values <- all(is.na(cleaned_data$units_arrived_at_scene))
paste("Any NA Values Test: ", any_NA_values)

#### Save data ####
write_csv(cleaned_data, "data/analysis_data/cleaned_ambulance_response_data.csv")

File: scripts/02-data_cleaning.py
#### Preamble ####
# Purpose: Cleans the raw ambulance dispatch response data
# Author: Muhammad Abdullah Motasim
# Date: 22 September 2024
# Contact: abdullah.motasim@mail.utoronto.ca
# License: MIT
# Pre-requisites: tidyverse, dplyr, raw_ambulance_response_data.csv
# Any other information needed? None

#### Workspace setup ####
import pandas as pd
import numpy as np

#### Clean data ####
raw_data = pd.read_csv("data/raw_data/raw_ambulance_response_data.csv")

# Select columns 2 to 5 (Python uses 0-based indexing, so 1:5)
data = raw_data.iloc[:, 1:5].copy()

# Drop missing rows
data = data.dropna()

# Split 'dispatch_time' into date and time
date_time_split = data["dispatch_time"].str.split(" ", expand=True)
data["date"] = date_time_split[0]
data["time"] = date_time_split[1]

# Split date into year, month, day
date_parts = data["date"].str.split("-", expand=True)
data["year"] = date_parts[0]
data["month"] = date_parts[1]
data["day"] = date_parts[2]

# Split time into hour, minute, second
time_parts = data["time"].str.split(":", expand=True)
data["hour"] = time_parts[0].fillna("00")
data["minute"] = time_parts[1].fillna("00")
data["second"] = time_parts[2].fillna("00")

# Create dispatch_time column as datetime
data["dispatch_time"] = pd.to_datetime(
    data["year"] + "-" + data["month"] + "-" + data["day"] + " " +
    data["hour"] + ":" + data["minute"] + ":" + data["second"],
    errors="coerce"
)

# Keep and reorder necessary columns
cleaned_data = data[[
    "dispatch_time",
    "incident_type",
    "priority_number",
    "units_arrived_at_scene",
    "year"
]]

#### Test data ####

# Test for no negative number of dispatchers
print("Negative Number of Dispatchers Test: ", cleaned_data["units_arrived_at_scene"].min() < 0)

# Test for dates within expected year range
years = ['2017', '2018', '2019', '2020', '2021', '2022']
print("Dispatch Time out-of-range Test: ", any(~cleaned_data["year"].isin(years)))

# Test for priority level meets requirements
priority_levels = [1, 3, 4, 5, 9, 11, 12, 13, 14]
print("Priority Level Outside of Range Test: ", any(~cleaned_data["priority_number"].isin(priority_levels)))

# Test for missing values
print("Any NA Values Test: ", cleaned_data.isna().any().any())

#### Save data ####
cleaned_data.to_csv("data/analysis_data/cleaned_ambulance_response_data.csv", index=False)
